internal/statemachine/helper.go

```diff
@@ -657,3 +658,133 @@ func generateDebootstrapCmd(imageDefinition ImageDefinition, targetDir string, i
 
 	return debootstrapCmd
 }
+
+// generateAptCmd generates the apt command used to create a chroot
+// environment that will eventually become the rootfs of the resulting image
+func generateAptCmds(targetDir string, packageList []string) []*exec.Cmd {
+	updateCmd := execCommand("chroot", targetDir, "apt", "update")
+
+	installCmd := execCommand("chroot", targetDir, "apt", "install",
+		"--assume-yes",
+		"--quiet",
+		"--option=Dpkg::options::=--force-unsafe-io",
+		"--option=Dpkg::Options::=--force-confold",
+	)
+
+	for _, aptPackage := range packageList {
+		installCmd.Args = append(installCmd.Args, aptPackage)
+	}
+
+	// Env is sometimes used for mocking command calls in tests,
+	// so only overwrite env if it is nil
+	if installCmd.Env == nil {
+		installCmd.Env = os.Environ()
+	}
+	installCmd.Env = append(installCmd.Env, "DEBIAN_FRONTEND=noninteractive")
+
+	return []*exec.Cmd{updateCmd, installCmd}
+}
+
+// createPPAInfo generates the name for a PPA sources.list file
+// in the convention of add-apt-repository, and the contents
+// that define the sources.list in the DEB822 format
+func createPPAInfo(ppa *PPAType, series string) (fileName string, fileContents string) {
+	splitName := strings.Split(ppa.PPAName, "/")
+	user := splitName[0]
+	ppaName := splitName[1]
+
+	fileName = fmt.Sprintf("%s-ubuntu-%s-%s.sources", user, ppaName, series)
+
+	var domain string
+	if ppa.Auth == "" {
+		domain = "https://ppa.launchpadcontent.net"
+	} else {
+		domain = fmt.Sprintf("https://%s@private-ppa.launchpadcontent.net", ppa.Auth)
+	}
+
+	fullDomain := fmt.Sprintf("%s/%s/%s/ubuntu", domain, user, ppaName)
+	fileContents = fmt.Sprintf("X-Repolib-Name: %s\nEnabled: yes\nTypes: deb\n"+
```

Comment:
This is good, but just mentioning something we *might* need to do in the future. So deb822 is 'the thing' and we should use it whenever possible, but I'm thinking about the need to do a 'backward compatibility layer' in case someone wants to build an older Ubuntu image. Do we know how far into the past does deb822 support in apt extend? Is it supported in bionic? Since eventually we'd want ubuntu-image to be able to build all still-supported series, and since we're running this in the target chroot, we might need to be considerate of the old times as well. I don't know if we should be aiming to support older series like xenial and trusty as those are basically ESM now, so not really image-building worthy.

Just something to keep in mind. Then again, this would be a good investigation, because we'd need to think how to add such compatibility levels and not risking bloating the codebase with multiple per-series hacks and changes. Since that's just one step away from becoming the "livecd-rootfs written in Go" :D

Reply:
I've got good news for you here! deb822 sources are supported as far back as trusty.